<?xml version="1.0" encoding="utf-8"?>
<!-- This URDF was automatically created by SolidWorks to URDF Exporter! Originally created by Stephen Brawner (brawner@gmail.com) 
     Commit Version: 1.6.0-1-g15f4949  Build Version: 1.6.7594.29634
     For more information, please see http://wiki.ros.org/sw_urdf_exporter -->
<robot

  name="wheel_leg" xmlns:xacro="http://www.ros.org/wiki/xacro">
  <!-- Import all Gazebo-customization elements, including Gazebo colors -->
  <xacro:include filename="$(find wheel_leg)/urdf/wheel_leg.gazebo" />
  
  <link name="base_footprint" />

  <joint name="joint_base_footprint_base_link" type="fixed">
  <origin
      xyz="0 0 0.3"
      rpy="0 0 0" />
    <parent link="base_footprint" />
    <child link="base_link" />
  </joint>

  <link
    name="base_link">
    <inertial>
      <origin
        xyz="-6.9389E-18 2.5278E-17 0"
        rpy="0 0 0" />
      <mass
        value="3.0" />
      <inertia
        ixx="0.0104"
        ixy="2.2956E-19"
        ixz="1.2678E-19"
        iyy="0.0035"
        iyz="-9.9634E-21"
        izz="0.0132" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://wheel_leg/meshes2/base_link.STL" /> 
      </geometry>
      <material
        name="">
        <color
          rgba="0.79216 0.81961 0.93333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <box size="0.12 0.200 0.040"/> 
      </geometry>
    </collision>
  </link>
  <!-- right -->
    <link
    name="link_right_hip">
    <inertial>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <mass
        value="0.01" />
      <inertia
        ixx="1E-5"
        ixy="0.0"
        ixz="0.0"
        iyy="1E-5"
        iyz="0.0"
        izz="1E-5" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <sphere radius = "0.005"/>
      </geometry>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <sphere radius = "0.005"/>
      </geometry>
    </collision>
  </link>
  <joint
    name="joint_base_right_hip"
    type="revolute">
    <origin
      xyz="0 -0.1 0"
      rpy="1.5708 1.5708 0" />
    <parent
      link="base_link" />
    <child
      link="link_right_hip" />
    <axis
      xyz="0 1 0" />
    <limit
      lower="-0.524"
      upper="1.7"
      effort="100"
      velocity="25" />
    <dynamics damping="0.7" friction="1.0"/>
  </joint>
  <link
    name="link_right_leg1">
    <inertial>
      <origin
        xyz="0.065 6.0715E-18 0.0038112"
        rpy="0 0 0" />
      <mass
        value="0.017303" />
      <inertia
        ixx="4.2828E-07"
        ixy="7.91E-23"
        ixz="-5.6741E-22"
        iyy="2.9677E-05"
        iyz="1.7088E-22"
        izz="2.9883E-05" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://wheel_leg/meshes2/link_right_leg1.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.79216 0.81961 0.93333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0.065 0.0 0.004"
        rpy="0 0 0" />
      <geometry>
        <box size="0.14 0.015 0.0080"/> 
      </geometry>
    </collision>
  </link>
  <joint
    name="joint_right_hip_right_leg1"
    type="revolute">
    <origin
      xyz="0 0 0"
      rpy="0 0 0" />
    <parent
      link="link_right_hip" />
    <child
      link="link_right_leg1" />
    <axis
      xyz="0 0 1" />
    <limit
      lower="-1.7"
      upper="1.7"
      effort="100"
      velocity="25" />
    <dynamics damping="0.7" friction="1.0"/>
  </joint>
  <link
    name="link_right_knee">
    <inertial>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <mass
        value="0.01" />
      <inertia
        ixx="1E-5"
        ixy="0.0"
        ixz="0.0"
        iyy="1E-5"
        iyz="0.0"
        izz="1E-5" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <sphere radius = "0.005"/>
      </geometry>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <sphere radius = "0.005"/>
      </geometry>
    </collision>
  </link>
  <joint
    name="joint_right_leg1_knee"
    type="revolute">
    <origin
      xyz="0.13 0 0"
      rpy="0 0 0" />
    <parent
      link="link_right_leg1" />
    <child
      link="link_right_knee" />
    <axis
      xyz="0 -1 0" />
    <limit
      lower="-1.7"
      upper="1.7"
      effort="100"
      velocity="25" />
    <dynamics damping="0.7" friction="1.0"/>
  </joint>
  <link
    name="link_right_leg2">
    <inertial>
      <origin
        xyz="0.067551 1.3878E-17 -0.0039019"
        rpy="0 0 0" />
      <mass
        value="0.016649" />
      <inertia
        ixx="4.1283E-07"
        ixy="-5.4594E-21"
        ixz="-1.0202E-07"
        iyy="2.6792E-05"
        iyz="7.6791E-23"
        izz="2.7008E-05" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://wheel_leg/meshes2/link_right_leg2.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.79216 0.81961 0.93333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0.065 0.0 -0.004"
        rpy="0 0 0" />
      <geometry>
        <box size="0.14 0.015 0.0080"/> 
      </geometry>
    </collision>
  </link>
  <joint
    name="joint_right_knee_leg2"
    type="revolute">
    <origin
      xyz="0.0 0 0"
      rpy="0 0 0" />
    <parent
      link="link_right_knee" />
    <child
      link="link_right_leg2" />
    <axis
      xyz="0 0 1" />
    <limit
      lower="-2.4"
      upper="2.4"
      effort="100"
      velocity="25" />
    <dynamics damping="0.7" friction="1.0"/>
  </joint>
  <link
    name="link_right_wheel">
    <inertial>
      <origin
        xyz="-1.7347E-18 0 0"
        rpy="0 0 0" />
      <mass
        value="0.09293" />
      <inertia
        ixx="3.8232E-05"
        ixy="-6.9397E-20"
        ixz="-4.8896E-23"
        iyy="7.0742E-05"
        iyz="4.4857E-20"
        izz="3.8232E-05" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://wheel_leg/meshes2/link_right_wheel.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.79216 0.81961 0.93333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="1.57 0 0" />
      <geometry>
        <cylinder length="0.02" radius="0.04"/>
      </geometry>
    </collision>
  </link>
  <joint
    name="joint_right_leg2_wheel"
    type="continuous">
    <origin
      xyz="0.13 0 0.01"
      rpy="-1.5708 0 1.5708" />
    <parent
      link="link_right_leg2" />
    <child
      link="link_right_wheel" />
    <axis
      xyz="0 1 0" />
    <dynamics damping="0.1" friction="0.0"/>
  </joint>
  <!-- left -->
  <link
    name="link_left_hip">
    <inertial>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <mass
        value="0.01" />
      <inertia
        ixx="1E-5"
        ixy="0.0"
        ixz="0.0"
        iyy="1E-5"
        iyz="0.0"
        izz="1E-5" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <sphere radius = "0.005"/>
      </geometry>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <sphere radius = "0.005"/>
      </geometry>
    </collision>
  </link>
  <joint
    name="joint_base_left_hip"
    type="revolute">
    <origin
      xyz="0 0.1 0"
      rpy="1.5708 1.5708 0" />
    <parent
      link="base_link" />
    <child
      link="link_left_hip" />
    <axis
      xyz="0 1 0" />
    <limit
      lower="-0.524"
      upper="1.7"
      effort="100"
      velocity="25" />
    <dynamics damping="0.7" friction="1.0"/>
  </joint>
  <link
    name="link_left_leg1">
    <inertial>
      <origin
        xyz="0.065 3.4694E-18 -0.0038112"
        rpy="0 0 0" />
      <mass
        value="0.017303" />
      <inertia
        ixx="4.2828E-07"
        ixy="1.7366E-22"
        ixz="2.8691E-22"
        iyy="2.9677E-05"
        iyz="1.8607E-23"
        izz="2.9883E-05" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://wheel_leg/meshes2/link_left_leg1.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.79216 0.81961 0.93333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0.065 0.0 -0.004"
        rpy="0 0 0" />
      <geometry>
        <box size="0.14 0.015 0.0080"/> 
      </geometry>
    </collision>
  </link>
  <joint
    name="joint_left_hip_left_leg1"
    type="revolute">
    <origin
      xyz="0 0 0"
      rpy="0 0 0" />
    <parent
      link="link_left_hip" />
    <child
      link="link_left_leg1" />
    <axis
      xyz="0 0 1" />
    <limit
      lower="-1.7"
      upper="1.7"
      effort="100"
      velocity="25" />
    <dynamics damping="0.7" friction="1.0"/>
  </joint>
  <link
    name="link_left_knee">
    <inertial>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <mass
        value="0.01" />
      <inertia
        ixx="1E-5"
        ixy="0.0"
        ixz="0.0"
        iyy="1E-5"
        iyz="0.0"
        izz="1E-5" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <sphere radius = "0.005"/>
      </geometry>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <sphere radius = "0.005"/>
      </geometry>
    </collision>
  </link>
  <joint
    name="joint_left_leg1_knee"
    type="revolute">
    <origin
      xyz="0.13 0 0"
      rpy="0 0 0" />
    <parent
      link="link_left_leg1" />
    <child
      link="link_left_knee" />
    <axis
      xyz="0 1 0" />
    <limit
      lower="-1.7"
      upper="1.7"
      effort="100"
      velocity="25" />
    <dynamics damping="0.7" friction="1.0"/>
  </joint>
  <link
    name="link_left_leg2">
    <inertial>
      <origin
        xyz="0.067551 -1.1276E-17 0.0039019"
        rpy="0 0 0" />
      <mass
        value="0.016649" />
      <inertia
        ixx="4.1283E-07"
        ixy="-2.4095E-22"
        ixz="1.0202E-07"
        iyy="2.6792E-05"
        iyz="-4.2458E-22"
        izz="2.7008E-05" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://wheel_leg/meshes2/link_left_leg2.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.79216 0.81961 0.93333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0.065 0.0 0.004"
        rpy="0 0 0" />
      <geometry>
        <box size="0.14 0.015 0.0080"/> 
      </geometry>
    </collision>
  </link>
  <joint
    name="joint_left_knee_leg2"
    type="revolute">
    <origin
      xyz="0.0 0 0"
      rpy="0 0 0" />
    <parent
      link="link_left_knee" />
    <child
      link="link_left_leg2" />
    <axis
      xyz="0 0 1" />
    <limit
      lower="-2.4"
      upper="2.4"
      effort="100"
      velocity="25" />
    <dynamics damping="0.7" friction="1.0"/>
  </joint>
  <link
    name="link_left_wheel">
    <inertial>
      <origin
        xyz="-8.6736E-19 0 0"
        rpy="0 0 0" />
      <mass
        value="0.09293" />
      <inertia
        ixx="3.8232E-05"
        ixy="-3.8909E-20"
        ixz="2.5573E-21"
        iyy="7.0742E-05"
        iyz="-5.4999E-21"
        izz="3.8232E-05" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://wheel_leg/meshes2/link_left_wheel.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.79216 0.81961 0.93333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="1.57 0 0" />
      <geometry>
        <cylinder length="0.02" radius="0.04"/>
      </geometry>
    </collision>
  </link>
  <joint
    name="joint_left_leg2_wheel"
    type="continuous">
    <origin
      xyz="0.13 0 -0.01"
      rpy="-1.5708 0 1.5708" />
    <parent
      link="link_left_leg2" />
    <child
      link="link_left_wheel" />
    <axis
      xyz="0 1 0" />
      <dynamics damping="0.1" friction="0.0"/>
  </joint>
  <link name="imu_link"/>
  <joint name="base_to_imu" type="fixed">
    <axis rpy="0 0 0" xyz="0 0 0"/>
    <parent link="base_link"/>
    <child link="imu_link"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
  </joint>


  <xacro:macro name="transmission_block" params="joint_name">
    <transmission name="trans1">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="${joint_name}">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
      <actuator name="motor1">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>
  </xacro:macro>

    <xacro:macro name="transmission_block2" params="joint_name">
    <transmission name="trans1">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="${joint_name}">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
      <actuator name="motor1">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        <mechanicalReduction>100</mechanicalReduction>
      </actuator>
    </transmission>
  </xacro:macro>

  <xacro:transmission_block2 joint_name="joint_base_right_hip"/>
  <xacro:transmission_block2 joint_name="joint_right_hip_right_leg1"/>
  <xacro:transmission_block2 joint_name="joint_base_left_hip"/>
  <xacro:transmission_block2 joint_name="joint_left_hip_left_leg1"/>
  <xacro:transmission_block2 joint_name="joint_right_leg1_knee"/>
  <xacro:transmission_block2 joint_name="joint_right_knee_leg2"/>
  <xacro:transmission_block2 joint_name="joint_left_leg1_knee"/>
  <xacro:transmission_block2 joint_name="joint_left_knee_leg2"/>
  <xacro:transmission_block2 joint_name="joint_left_leg2_wheel"/>
  <xacro:transmission_block joint_name="joint_right_leg2_wheel"/>
</robot>